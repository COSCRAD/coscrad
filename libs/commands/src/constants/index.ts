export const Ack = Symbol('The command has succeeded');

export type Ack = typeof Ack;
