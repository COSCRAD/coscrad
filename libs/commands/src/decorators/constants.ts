export const COMMAND_HANDLER_METADATA = '__commandHandler_';

export const COMMAND_METADATA = '__command__';
