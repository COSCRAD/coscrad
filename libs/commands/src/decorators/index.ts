export * from './command-handler.decorator';
export * from './command.decorator';
export * from './types';
