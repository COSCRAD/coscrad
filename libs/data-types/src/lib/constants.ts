export const COSCRAD_DATA_TYPE_METADATA = '__coscrad-data-types__';

export const COSCRAD_UNION_MEMBER_METADATA = '__coscrad-union-member__';
