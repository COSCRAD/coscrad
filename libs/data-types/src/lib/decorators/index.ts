export * from './CompositeIdentifier';
export * from './Enum';
export * from './NestedDataType';
export * from './NonEmptyString';
export * from './NonNegativeFiniteNumber';
export * from './RawData';
export * from './URL';
export * from './UUID';
