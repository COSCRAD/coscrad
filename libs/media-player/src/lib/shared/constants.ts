export const EDITOR_SOUND_BAR_HEIGHT_IN_PIXELS = 40;

export const RULER_HEIGHT_IN_PIXELS = 30;

export const RULER_TICK_WIDTH_IN_PIXELS = 1;

export const RULER_TICK_HEIGHT_IN_PIXELS = 10;

export const RULER_TICK_WITH_NUMBER_HEIGHT_IN_PIXELS = 20;

export const RULER_TICKS_AND_NUMBERS_COLOR = '#178e06';

/**
 * This limitation is from the IE 9 / 10 (Win) implementation of Canvas
 * https://jhildenbiddle.github.io/canvas-size/#/?id=test-results
 */
export const MAX_CANVAS_WIDTH_IN_PIXELS = 8000;

export type ZoomLevels = {
    zoomLevel: number;
    rulerTickXCoordinateOffset: number;
    rulerTickFrequencyInSeconds: number;
    timecodeDisplayFrequencyInSeconds: number;
};

// re-order zoom levels - 0 is deep zoom, highest is most zoomed out
// How to dynamically zoom to level where whole timeline is in view
export const ZOOM_LEVELS: ZoomLevels[] = [
    // 0
    {
        zoomLevel: 0,
        rulerTickXCoordinateOffset: 40,
        rulerTickFrequencyInSeconds: 1,
        timecodeDisplayFrequencyInSeconds: 10,
    },
    // 1
    {
        zoomLevel: 1,
        rulerTickXCoordinateOffset: 20,
        rulerTickFrequencyInSeconds: 1,
        timecodeDisplayFrequencyInSeconds: 10,
    },
    // 2
    {
        zoomLevel: 2,
        rulerTickXCoordinateOffset: 10,
        rulerTickFrequencyInSeconds: 1,
        timecodeDisplayFrequencyInSeconds: 10,
    },
    // 3
    {
        zoomLevel: 3,
        rulerTickXCoordinateOffset: 40,
        rulerTickFrequencyInSeconds: 5,
        timecodeDisplayFrequencyInSeconds: 10,
    },
    // 4
    {
        zoomLevel: 4,
        rulerTickXCoordinateOffset: 10,
        rulerTickFrequencyInSeconds: 1,
        timecodeDisplayFrequencyInSeconds: 20,
    },
    // 5
    {
        zoomLevel: 5,
        rulerTickXCoordinateOffset: 40,
        rulerTickFrequencyInSeconds: 5,
        timecodeDisplayFrequencyInSeconds: 10,
    },
];
