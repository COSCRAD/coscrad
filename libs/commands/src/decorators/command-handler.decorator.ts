import { Type } from '@nestjs/common';
import { ICommand } from '../interfaces/command.interface';
import { COMMAND_HANDLER_METADATA } from './constants';

/**
 * This decorator marks a class as a Command Handler.
 *
 * @param command command (constructor) that is handled by this command handler
 */
export const CommandHandler =
    (command: Type<ICommand>): ClassDecorator =>
    (target: object) => {
        Reflect.defineMetadata(COMMAND_HANDLER_METADATA, { command }, target);
    };
