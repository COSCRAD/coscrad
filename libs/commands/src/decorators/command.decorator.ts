import { EmptyCommandTypeException } from '../exceptions/empty-command-type-exception';
import { COMMAND_METADATA } from './constants';
import { CommandMetadataBase } from './types/CommandMetadataBase';

/**
 * This decorator marks a class as a Command.
 *
 * @param metadata an object that includes
 * - a `type` property - name of the command (e.g. 'ADD_WIDGET')
 * - any additional metadata (general commadn info) you need, pershaps to share
 * with the client through your API
 *
 */
export const Command =
    <TCommandMeta extends CommandMetadataBase>(metadata: TCommandMeta): ClassDecorator =>
    (target: object) => {
        const { type } = metadata;

        if (type.length === 0) {
            throw new EmptyCommandTypeException();
        }

        Reflect.defineMetadata(COMMAND_METADATA, metadata, target);
    };
