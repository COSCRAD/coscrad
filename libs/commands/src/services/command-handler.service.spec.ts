import { Test, TestingModule } from '@nestjs/testing';
import { Ack } from '../constants';
import { Command, CommandHandler } from '../decorators';
import { ICommandHandler } from '../interfaces/command-handler.interface';
import { ICommand } from '../interfaces/command.interface';
import { CommandHandlerService } from './command-handler.service';

describe('CommandsService', () => {
    let service: CommandHandlerService;

    /**
     * Note that only the `type` property is required from the point of view of
     * the `@coscrads/commands` lib. However additional metadata can be added as
     * needed by the client, and this will be returned from our
     * `CommandHandlerService` on request.
     */
    @Command({ type: 'ADD_WIDGET', label: 'Add Widget', info: 'Adds a Widget' })
    class AddWidget implements ICommand {
        public readonly widgetName: string;
    }

    @CommandHandler(AddWidget)
    class HandleAddWidget implements ICommandHandler {
        async execute({ widgetName }: AddWidget): Promise<Ack | Error> {
            return widgetName === 'fail' ? new Error('Add Widet failed') : Ack;
        }
    }

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            providers: [CommandHandlerService, AddWidget, HandleAddWidget],
        }).compile();

        service = module.get<CommandHandlerService>(CommandHandlerService);

        // This is normally done by the finder service, but there's a separate test for that
        service.registerHandler('ADD_WIDGET', new HandleAddWidget());
    });

    it('should be defined', () => {
        expect(service).toBeDefined();
    });

    describe('CommandHandlerService.getAllCommandCtorsAndMetadata', () => {
        it('should return the expected meta', () => {
            const result = service.getAllCommandCtorsAndMetadata();

            expect(result).toMatchSnapshot();

            expect(result[0].constructor).toBe(AddWidget);
        });
    });

    describe('CommandHandlerService.execute', () => {
        describe('when the command it receives is valid', () => {
            it('should return Ack', async () => {
                const result = await service.execute({
                    type: 'ADD_WIDGET',
                    payload: { widgetName: 'ok name' },
                });

                expect(result).toBe(Ack);
            });
        });

        describe('when the command it receives is invalid', () => {
            it('should return an error', async () => {
                const result = await service.execute({
                    type: 'ADD_WIDGET',
                    payload: { widgetName: 'fail' },
                });

                expect(result).toBeInstanceOf(Error);
            });
        });
    });
});
