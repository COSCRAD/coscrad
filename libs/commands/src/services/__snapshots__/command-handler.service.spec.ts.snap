// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandsService CommandHandlerService.getAllCommandCtorsAndMetadata should return the expected meta 1`] = `
[
  AddWidget {
    "constructor": [Function],
    "meta": {
      "info": "Adds a Widget",
      "label": "Add Widget",
      "type": "ADD_WIDGET",
    },
  },
]
`;
